src/synchronization_c.c4

```diff
@@ -147,20 +178,222 @@ SHMEM_BIND_C_WAIT(`SHMEM_DEF_WAIT')
+        free(completed);                                                                       \
+    }
+
+SHMEM_BIND_C_SYNC(`SHMEM_DEF_WAIT_UNTIL_ALL')
+
+#define SHMEM_DEF_WAIT_UNTIL_ANY(STYPE,TYPE)                                                   \
+    size_t SHMEM_FUNCTION_ATTRIBUTES                                                           \
+    shmemx_##STYPE##_wait_until_any(TYPE *vars, size_t nelems,                                 \
+                                    int * restrict status, int cond, TYPE value)               \
+    {                                                                                          \
+        SHMEM_ERR_CHECK_INITIALIZED();                                                         \
+        SHMEM_ERR_CHECK_SYMMETRIC(vars, sizeof(TYPE));                                         \
+        SHMEM_ERR_CHECK_CMP_OP(cond);                                                          \
+                                                                                               \
+        size_t ret = 0, i = 0, npasses = 0, no_wait_len = 0;                                   \
+        while (!ret) {                                                                         \
+            for (i = 0; i < nelems; i++) {                                                     \
+                if (status) {                                                                  \
+                    if (!status[i]) {                                                          \
+                        COMP(cond, vars[i], value, ret);                                       \
+                    } else {                                                                   \
+                        no_wait_len++;                                                         \
+                    }                                                                          \
+                } else {                                                                       \
+                    COMP(cond, vars[i], value, ret);                                           \
+                }                                                                              \
+                if (ret) {                                                                     \
+                    if (status) status[i] = 1;                                                 \
+                    break;                                                                     \
+                }                                                                              \
+            }                                                                                  \
+            if (no_wait_len == nelems && npasses == 0) {                                       \
+                ret = SIZE_MAX;                                                                \
+            }                                                                                  \
```

Comment:
Oh wow, this is complicated.  I would probably just check the `SIZE_MAX` cases before entering this loop.  I think you also missed the case where `nelems == 0`?